File: python/kungfu/wingchun/EventLoop.py

from kungfu.wingchun.TaskScheduler import TaskScheduler
from kungfu.wingchun.constants import MsgType
import nnpy, json, pyyjj
import sys
import signal

class EventLoop:
    def __init__(self, name):
        self._quit = False
        self._name = name
        self._reader = None
        self._scheduler = TaskScheduler()
        self._sockets = []
        self._quote_cb = None
        self._entrust_cb = None
        self._trans_cb = None
        self._login_cb = None
        self._sub_cb = None
        self._input_cb = None
        self._action_cb = None
        self._order_cb = None
        self._trade_cb = None
        self._algo_input_cb = None
        self._manual_order_input_cb = None
        self._manual_order_action_cb = None
        self._reload_instruments_cb = None
        self._signal_received = -1
        self._signal_callbacks = []

    def subscribe_nanomsg(self, url):
        socket = nnpy.Socket(nnpy.AF_SP, nnpy.SUB)
        socket.setsockopt(nnpy.SUB, nnpy.SUB_SUBSCRIBE, '')
        socket.connect(url)
        self._sockets.append(socket)

    def bind_nanomsg(self, url):
        socket = nnpy.Socket(nnpy.AF_SP, nnpy.REP)
        socket.bind(url)
        self._sockets.append(socket)

    def add_socket(self, socket):
        self._sockets.append(socket)

    def subscribe_yjj_journal(self, journal_folder, journal_name, offset_nano):
        if self._reader is None:
            self._reader = pyyjj.create_reader([journal_folder], [journal_name], offset_nano, self._name)
        else:
            idx = self._reader.add_journal(journal_folder, journal_name)
            self._reader.seek_time(idx, offset_nano)

    def get_nano(self):
        nano = self._scheduler.get_nano()
        if nano == 0:
            nano = pyyjj.nano_time()
        return nano

    def register_nanotime_callback(self, nano, callback):
        self._scheduler.insert_callback_at(nano, callback)

    def register_nanotime_callback_at_next(self, time_str, callback):
        self._scheduler.insert_callback_at_next(time_str, callback)

    def register_quote_callback(self, callback):
        self._quote_cb = callback

    def register_entrust_callback(self, callback):
        self._entrust_cb = callback

    def register_transaction_callback(self, callback):
        self._trans_cb = callback

    def register_login_callback(self, callback):
        self._login_cb = callback

    def register_subscribe_callback(self, callback):
        self._sub_cb = callback

    def register_order_input_callback(self, callback):
        self._input_cb = callback

    def register_order_action_callback(self, callback):
        self._action_cb = callback

    def register_order_callback(self, callback):
        self._order_cb = callback

    def register_trade_callback(self, callback):
        self._trade_cb = callback

    def register_algo_order_input_callback(self, callback):
        self._algo_input_cb = callback

    def register_manual_order_input_callback(self, callback):
        self._manual_order_input_cb = callback

    def register_manual_order_action_callback(self, callback):
        self._manual_order_action_cb = callback

    def register_reload_instruments_callback(self, callback):
        self._reload_instruments_cb = callback

    def register_signal_callback(self, callback):
        self._signal_callbacks.append(callback)

    def run(self):
        signal.signal(signal.SIGINT, self.__handle_signal)
        signal.signal(signal.SIGTERM, self.__handle_signal)
        #signal.signal(signal.SIGHUP, self.__handle_signal)
        #signal.signal(signal.SIGQUIT, self.__handle_signal)

        self._quit = False
        while not self._quit and self._signal_received < 0:
            self.__iteration()
        if self._signal_received >=0 :
            print("signal received ", self._signal_received)
            for cb in self._signal_callbacks:
                cb(self._signal_received)

    def stop(self):
        self._quit = True

    def __iteration(self):
        nano = -1
        if self._reader is not None:
            frame = self._reader.next()
            if frame is not None:
                nano = frame.nano()
                self.__handle_frame(frame)
        if nano == -1:
            nano = pyyjj.nano_time()
        self._scheduler.update_nano(nano)
        for socket in self._sockets:
            try:
                reply = socket.recv(nnpy.DONTWAIT)
                j_reply = json.loads(reply.decode('utf-8')[:-1])
                # compatible with req
                if "req" in j_reply and "msg_type" not in j_reply:
                    j_reply["msg_type"] = j_reply["req"]
                msg_type = j_reply["msg_type"]
                data = j_reply["data"]
                self.__handle_nanomsg(msg_type, data)
            except nnpy.NNError as nnpy_e:
                # if nothing received, nnpy will throw exception, wtf!!!!!
                pass
            except Exception as e:
                print(e)

    def __handle_frame(self, frame):
        msg_type = frame.msg_type()
        if msg_type == MsgType.Quote:
            if self._quote_cb is not None:
                self._quote_cb(frame.get_data())
        elif msg_type == MsgType.Entrust:
            if self._entrust_cb is not None:
                self._entrust_cb(frame.get_data())
        elif msg_type == MsgType.Transaction:
            if self._trans_cb is not None:
                self._trans_cb(frame.get_data())
        elif msg_type == MsgType.OrderInput:
            if self._input_cb is not None:
                self._input_cb(frame.get_data())
        elif msg_type == MsgType.OrderAction:
            if self._action_cb is not None:
                self._action_cb(frame.get_data())
        elif msg_type == MsgType.Order:
            if self._order_cb is not None:
                self._order_cb(frame.get_data())
        elif msg_type == MsgType.Trade:
            if self._trade_cb is not None:
                self._trade_cb(frame.get_data())
        elif msg_type == MsgType.AlgoOrderInput:
            if self._algo_input_cb is not None:
                try:
                    j = json.loads(frame.get_data())
                    self._algo_input_cb(j['order_id'], j['algo_type'], j['client_id'], j['algo_order_input'])
                except Exception as e:
                    print(e)
        else:
            pass

    def __handle_nanomsg(self, msg_type, data):
        if msg_type == MsgType.ReqLogin:
            if self._login_cb is not None:
                recipient = data["recipient"]
                sender = data["sender"]
                self._login_cb(recipient, sender)
        elif msg_type == MsgType.Subscribe:
            if self._sub_cb is not None:
                recipient = data["recipient"]
                instruments = data["instruments"]
                is_level2 = data["is_level2"]
                self._sub_cb(recipient, instruments, is_level2)
        elif msg_type == MsgType.ReloadFutureInstrument:
            if self._reload_instruments_cb is not None:
                self._reload_instruments_cb()
        elif msg_type == MsgType.ReqOrderAction:
            if self._manual_order_action_cb is not None:
                account_id = ''
                if "account_id" in data:
                    account_id = data["account_id"]
                client_id = ''
                if "client_id" in data:
                    client_id = data["client_id"]
                order_ids = []
                if "order_id" in data:
                    for item in data["order_id"]:
                        order_ids.append(int(item))
                self._manual_order_action_cb(account_id, client_id, order_ids)
        else:
            pass

    def __handle_signal(self, signal, frame):
        self._signal_received = signal

File: cpp/wingchun/src/event_loop/event_loop.cpp
//
// Created by qlu on 2019/1/16.
//
#include <nlohmann/json.hpp>
#include <nanomsg/pubsub.h>
#include <nanomsg/reqrep.h>

#include <kungfu/time/timer.h>
#include <kungfu/wingchun/msg.h>
#include <kungfu/wingchun/event_loop/event_loop.h>
#include <kungfu/wingchun/reqrsp.h>
#include <kungfu/wingchun/serialize.h>
#include <kungfu/wingchun/oms/def.h>

using namespace kungfu::journal;
using namespace kungfu::flying;

namespace kungfu
{
    volatile sig_atomic_t EventLoop::signal_received_ = -1;

    void EventLoop::set_logger(std::shared_ptr<spdlog::logger> logger) const
    {
        spdlog::set_default_logger(logger->clone("event_loop"));
    };

    void EventLoop::subscribe_nanomsg(const std::string& url)
    {
        std::shared_ptr<yijinjing::nanomsg::socket> socket = std::shared_ptr<yijinjing::nanomsg::socket>(new yijinjing::nanomsg::socket(AF_SP, NN_SUB));
        socket->connect(url.c_str());
        socket->setsockopt(NN_SUB, NN_SUB_SUBSCRIBE, "", 0);
        socket_vec_.emplace_back(socket);
    }

    void EventLoop::bind_nanomsg(const std::string &url)
    {
        std::shared_ptr<yijinjing::nanomsg::socket> socket = std::shared_ptr<yijinjing::nanomsg::socket>(new yijinjing::nanomsg::socket(AF_SP, NN_REP));
        socket->bind(url.c_str());
        socket_vec_.emplace_back(socket);
    }

    void EventLoop::subscribe_yjj_journal(const std::string& journal_folder, const std::string& journal_name, int64_t offset_nano)
    {
        if (reader_.get() == nullptr)
        {
            reader_ = kungfu::yijinjing::JournalReader::create(journal_folder, journal_name, offset_nano, name_);
        }
        else
        {
            int journal_idx = reader_->addJournal(journal_folder, journal_name);
            reader_->seekTimeJournal(journal_idx, offset_nano);
        }
    }

    void EventLoop::register_nanotime_callback(int64_t nano, TSCallback callback)
    {
        scheduler_->insert_callback_at(nano, callback);
    }

    void EventLoop::register_nanotime_callback_at_next(const char *time_str, kungfu::TSCallback callback)
    {
        scheduler_->insert_callback_at_next(time_str, callback);
    }

    void EventLoop::register_quote_callback(QuoteCallback callback)
    {
        quote_callback_ = callback;
    }

    void EventLoop::register_entrust_callback(EntrustCallback callback)
    {
        entrust_callback_ = callback;
    }

    void EventLoop::register_transaction_callback(TransactionCallback callback)
    {
        transaction_callback_ = callback;
    }

    void EventLoop::register_req_login_callback(ReqLoginCallback callback)
    {
        login_callback_ = callback;
    }

    void EventLoop::register_subscribe_callback(SubscribeCallback callback)
    {
        sub_callback_ = callback;
    }

    void EventLoop::register_order_input_callback(OrderInputCallback callback)
    {
        order_input_callback_ = callback;
    }

    void EventLoop::register_order_action_callback(OrderActionCallback callback)
    {
        order_action_callback_ = callback;
    }

    void EventLoop::register_order_callback(OrderCallback callback)
    {
        order_callback_ = callback;
    }

    void EventLoop::register_trade_callback(TradeCallback callback)
    {
        trade_callback_ = callback;
    }

    void EventLoop::register_algo_order_input_callback(AlgoOrderInputCallback callback)
    {
        algo_order_input_callback_ = callback;
    }

    void EventLoop::register_algo_order_status_callback(AlgoOrderStatusCallback callback)
    {
        algo_order_status_callback_ = callback;
    }

    void EventLoop::register_algo_order_action_callback(AlgoOrderActionCallback callback)
    {
        algo_order_action_callback_ = callback;
    }

    void EventLoop::register_manual_order_input_callback(kungfu::ManualOrderInputCallback callback)
    {
        manual_order_input_callback_ = callback;
    }

    void EventLoop::register_manual_order_action_callback(kungfu::ManualOrderActionCallback callback)
    {
        manual_order_action_callback_ = callback;
    }

    void EventLoop::register_reload_instruments_callback(kungfu::ReloadInstrumentsCallback callback)
    {
        reload_instruments_callback_ = callback;
    }

    void EventLoop::register_signal_callback(SignalCallback signal_callback)
    {
        signal_callbacks_.push_back(signal_callback);
    }

    void EventLoop::signal_handler(int signal)
    {
        signal_received_ = signal;
    }

    void EventLoop::run()
    {
        signal(SIGINT, EventLoop::signal_handler);
        signal(SIGTERM, EventLoop::signal_handler);
#ifndef _WINDOWS
        signal(SIGHUP,  EventLoop::signal_handler);
        signal(SIGQUIT, EventLoop::signal_handler);
        signal(SIGKILL, EventLoop::signal_handler);
#endif

        quit_ = false;
        while (! quit_ && signal_received_ < 0)
        {
            iteration();
        }
        if (signal_received_ >= 0)
        {
            SPDLOG_TRACE("signal received {}", signal_received_);
            for (const auto& cb: signal_callbacks_)
            {
                cb(signal_received_);
            }
        }
    }

    void EventLoop::stop()
    {
        quit_= true;
    }

    void EventLoop::iteration()
    {
        if (!low_latency_ && notice_.wait())
        {
            notice_.last_message();
        }
        
        int64_t nano = -1;
        if (reader_ != nullptr)
        {
            yijinjing::FramePtr frame = reader_->getNextFrame();
            if (frame != nullptr)
            {
                nano = frame->getNano();
                int msg_type = frame->getMsgType();
                switch (msg_type)
                {
                    case (int)MsgType::Quote:
                    {
                        Quote* quote_ptr = (Quote*) frame->getData();
                        if (quote_callback_)
                        {
                            quote_callback_(*quote_ptr);
                        }
                        break;
                    }
                    case (int)MsgType::Entrust:
                    {
                        Entrust* entrust_ptr = (Entrust*) frame->getData();
                        if (entrust_callback_)
                        {
                            entrust_callback_(*entrust_ptr);
                        }
                        break;
                    }
                    case (int)MsgType::Transaction:
                    {
                        Transaction* transaction_ptr = (Transaction*) frame->getData();
                        if (transaction_callback_)
                        {
                            transaction_callback_(*transaction_ptr);
                        }
                        break;
                    }
                    case (int) MsgType::OrderInput:
                    {
                        OrderInput* input = (OrderInput*) frame->getData();
                        if (order_input_callback_)
                        {
                            order_input_callback_(*input);
                        }
                        break;
                    }
                    case (int)MsgType::OrderAction:
                    {
                        OrderAction* action = (OrderAction*) frame->getData();
                        if (order_action_callback_)
                        {
                            order_action_callback_(*action);
                        }
                        break;
                    }
                    case (int)MsgType::Order:
                    {
                        Order* order = (Order*) frame->getData();
                        if (order_callback_)
                        {
                            order_callback_(*order);
                        }
                        break;
                    }
                    case (int)MsgType::Trade:
                    {
                        Trade* trade = (Trade*) frame->getData();
                        if (trade_callback_)
                        {
                            trade_callback_(*trade);
                        }
                        break;
                    }
                    case (int)MsgType::AlgoOrderInput:
                    {
                        std::string js((char *) frame->getData());
                        try
                        {
                            nlohmann::json j = nlohmann::json::parse(js);
                            AlgoOrderInput msg = j;
                            if(algo_order_input_callback_)
                            {
                                algo_order_input_callback_(msg.order_id, msg.client_id, msg.algo_type, msg.input);
                            }
                        }
                        catch (std::exception& e)
                        {
                            SPDLOG_ERROR("failed to parse algo order input msg, data[{}], exception: {}", (char*)frame->getData(), e.what());
                        }
                        break;
                    }
                    case (int)MsgType::AlgoOrderStatus:
                    {
                        std::string js((char*) frame->getData());
                        try
                        {
                            nlohmann::json j = nlohmann::json::parse(js);
                            AlgoOrderStatus msg = j;
                            if(algo_order_status_callback_)
                            {
                                algo_order_status_callback_(msg.order_id, msg.algo_type, msg.status);
                            }
                        }
                        catch (std::exception& e)
                        {
                            SPDLOG_ERROR("failed to parse algo order action msg, data[{}], exception: {}", (char*)frame->getData(), e.what());
                        }
                        break;
                    }
                    case (int)MsgType::AlgoOrderAction:
                    {
                        std::string js((char*) frame->getData());
                        try
                        {
                            nlohmann::json j = nlohmann::json::parse(js);
                            AlgoOrderAction msg = j;
                            if(algo_order_action_callback_)
                            {
                                algo_order_action_callback_(msg.order_id, msg.order_action_id, msg.action);
                            }
                        }
                        catch (std::exception& e)
                        {
                            SPDLOG_ERROR("failed to parse algo order action msg, data[{}], exception: {}", (char*)frame->getData(), e.what());
                        }
                        break;
                    }
                    default:
                    {
                        break;
                    }
                }
            }
        }

        if (nano == -1)
        {
            nano = yijinjing::getNanoTime();
        }
        scheduler_->update_nano(nano);

        for (const auto& socket: socket_vec_)
        {
            int rc = socket->recv(NN_DONTWAIT);
            if (rc > 0)
            {
                SPDLOG_TRACE("recv: data[{}]\n", socket->last_message());
                try
                {
                    nlohmann::json content = nlohmann::json::parse(socket->last_message());
                    // 兼容req
                    if (content.find("req") != content.end() && content.find("msg_type") == content.end())
                    {
                        content["msg_type"] = content["req"];
                    }
                    NNMsg msg = content;
                    switch (msg.msg_type)
                    {
                        case MsgType::ReqLogin:
                        {
                            if (login_callback_)
                            {
                                LoginRequest req = msg.data;
                                login_callback_(req.recipient, req.sender);
                            }
                            break;
                        }
                        case MsgType::Subscribe:
                        {
                            if (sub_callback_)
                            {
                                SubscribeRequest req = msg.data;
                                sub_callback_(req.recipient, req.instruments, req.is_level2);
                            }
                            break;
                        }
                        case MsgType::OrderInput:
                        {
                            if (order_input_callback_)
                            {
                                OrderInput order_input = msg.data;
                                order_input_callback_(order_input);
                            }
                            break;
                        }
                        case MsgType::OrderAction:
                        {
                            if (order_action_callback_)
                            {
                                OrderAction order_action = msg.data;
                                order_action_callback_(order_action);
                            }
                            break;
                        }
                        case MsgType::ReloadFutureInstrument:
                        {
                            if (reload_instruments_callback_)
                            {
                                reload_instruments_callback_();
                            }
                            break;
                        }
                        case MsgType::ReqOrderInput:
                        {
                            if (manual_order_input_callback_)
                            {
                                OrderInput order_input = {};
                                strcpy(order_input.instrument_id, msg.data["instrument_id"].get<std::string>().c_str());
                                strcpy(order_input.account_id, msg.data["account_id"].get<std::string>().c_str());
                                strcpy(order_input.client_id, msg.data["client_id"].get<std::string>().c_str());
                                order_input.price_type = msg.data["price_type"].get<std::string>()[0];
                                order_input.limit_price = msg.data["limit_price"].get<double>();
                                order_input.volume = msg.data["volume"].get<int64_t>();
                                order_input.side = msg.data["side"].get<std::string>()[0];
                                order_input.offset = msg.data["offset"].get<std::string>()[0];

                                manual_order_input_callback_(order_input);
                            }
                        }
                        case MsgType::ReqOrderAction:
                        {
                            if (manual_order_action_callback_)
                            {
                                std::string account_id = "";
                                if (msg.data.find("account_id") != msg.data.end())
                                {
                                    account_id = msg.data["account_id"].get<std::string>();
                                }
                                std::string client_id = "";
                                if (msg.data.find("client_id") != msg.data.end())
                                {
                                    client_id = msg.data["client_id"].get<std::string>();
                                }
                                std::vector<uint64_t> order_ids;
                                if (msg.data.find("order_id") != msg.data.end())
                                {
                                    const auto& j_orders_ids = msg.data["order_id"];
                                    if (j_orders_ids.is_array())
                                    {
                                        for (auto i = 0; i < j_orders_ids.size(); ++i)
                                        {
                                            order_ids.emplace_back(std::stoull(j_orders_ids[i].get<std::string>()));
                                        }
                                    }
                                }
                                manual_order_action_callback_(account_id, client_id, order_ids);
                            }
                            break;
                        }
                        default:
                        {
                            break;
                        }
                    }
                }
                catch(std::exception &e)
                {
                    SPDLOG_ERROR("failed to parse data[{}], exception: {}", socket->last_message(), e.what());
                    continue;
                }
            }
        }
    }

    int64_t EventLoop::get_nano() const
    {
        int64_t nano = scheduler_->get_nano();
        return nano > 0 ? nano : kungfu::yijinjing::getNanoTime();
    }
}
